app/models/report.rb

```diff
@@ -4,14 +4,32 @@ class Report < ApplicationRecord
   belongs_to :user
   has_many :comments, as: :commentable, dependent: :destroy
 
+  has_many :active_mentions, class_name: 'ReportMention', foreign_key: :mentioned_by_id, dependent: :destroy, inverse_of: :mentioned_by
+  has_many :mentioning_reports, through: :active_mentions, source: :mention_to
+
+  has_many :passive_mentions, class_name: 'ReportMention', foreign_key: :mention_to_id, dependent: :destroy, inverse_of: :mention_to
+  has_many :mentioned_reports, through: :passive_mentions, source: :mentioned_by
+
   validates :title, presence: true
   validates :content, presence: true
 
+  after_save :save_mentions
+
   def editable?(target_user)
     user == target_user
   end
 
   def created_on
     created_at.to_date
   end
+
+  private
+
+  MENTION_REGEXP = %r{http://localhost:3000/reports/(\d+)}
```

Comment:
（別解）後読みを使うと31行目の`flatten`をなくせます。

```suggestion
  MENTION_REGEXP = %r{(?<=http://localhost:3000/reports/)\d+}
```